src/dune_rules/coq_rules.ml

```diff
@@ -265,71 +265,82 @@ module Context = struct
-    let deps = String.extract_blank_separated_words deps in
-    if coq_debug then
-      Format.eprintf "deps for %s: %a@\n%!"
-        (Path.Build.to_string source)
-        (Format.pp_print_list Format.pp_print_string)
-        deps;
-    (* Add prelude deps for when stdlib is in scope and we are not actually
-       compiling the prelude *)
-    let deps = List.map ~f:(Path.relative (Path.build dir)) deps in
-    match boot_type with
-    | No_boot
-    | Bootstrap_prelude ->
-      deps
-    | Bootstrap lib ->
-      Path.relative (Path.build (Coq_lib.src_root lib)) "Init/Prelude.vo"
-      :: deps )
+  let for_ ext =
+    match
+      List.find_map deps ~f:(fun (targets, deps) ->
+          List.find_map targets ~f:(fun t ->
+              if Filename.check_suffix t ext then
+                Some (t, deps)
+              else
+                None))
+    with
+    | None -> invalid ".vo not found"
+    | Some (basename, deps) -> (
+      let ff = List.hd @@ String.extract_blank_separated_words basename in
+      let depname, _ = Filename.split_extension ff in
+      let modname =
+        String.concat ~sep:"/"
+          Coq_module.(
+            prefix coq_module @ [ Coq_module.Name.to_string (name coq_module) ])
+      in
+      if coq_debug then
+        Format.eprintf "depname / modname: %s / %s@\n%!" depname modname;
+      if depname <> modname then invalid "basename";
+      if coq_debug then
+        Format.eprintf "deps for %s: %a@\n%!"
+          (Path.Build.to_string source)
+          (Format.pp_print_list Format.pp_print_string)
+          deps;
+      (* Add prelude deps for when stdlib is in scope and we are not actually
+         compiling the prelude *)
+      let deps = List.map ~f:(Path.relative (Path.build dir)) deps in
+      match boot_type with
+      | No_boot
+      | Bootstrap_prelude ->
+        deps
+      | Bootstrap lib ->
+        Path.relative (Path.build (Coq_lib.src_root lib)) ("Init/Prelude" ^ ext)
+        :: deps )
+  in
+  (for_ ".vo", for_ ".vos")
 
 let deps_of ~dir ~boot_type coq_module =
   let stdout_to = Coq_module.dep_file ~obj_dir:dir coq_module in
-  Action_builder.dyn_paths_unit
-    (Action_builder.map
-       (Action_builder.lines_of (Path.build stdout_to))
-       ~f:(parse_coqdep ~dir ~boot_type ~coq_module))
+  let deps =
+    Action_builder.memoize
+      (Path.Build.to_string stdout_to)
+      (Action_builder.map
+         (Action_builder.lines_of (Path.build stdout_to))
+         ~f:(parse_coqdep ~dir ~boot_type ~coq_module))
+  in
+  let deps f = Action_builder.map deps ~f |> Action_builder.dyn_paths_unit in
+  (deps fst, deps snd)
 
 let coqdep_rule (cctx : _ Context.t) ~source_rule ~file_flags coq_module =
   (* coqdep needs the full source + plugin's mlpack to be present :( *)
   let source = Coq_module.source coq_module in
   let file_flags =
     [ Command.Args.S file_flags
-    ; As [ "-dyndep"; "opt" ]
+    ; As [ "-dyndep"; "opt"; "-vos" ]
```

Comment:
This needs to be conditional as to support Coq older than 8.11

Reply:
But we don't have a way to check the version of Coq. Perhaps it should be enabled explicitly in the project file? E.g. `(coq.vos enable)`

Reply:
We can also say that `(coq lang 0.4)` is required and only 8.11 upwards is supported